### DISCUSSION TITLE
Using JavaScript to communicate with Pure Data

### DISCUSSION THREAD

**Discussion by luismrguimaraes** [charlesneimog/pd4web#16] [Q&A 🇺🇸] [answered]
Is there a way to do this? In my case, I want to use Pure Data as the audio engine of a Three JS game, but was unable to figure out how to communicate with the patch from JavaScript.

**Comment by charlesneimog** [comment, marked as answer]
Hi @luismrguimaraes, 

Yes, it is possible I am yet doing the docs, but you can understand part of it reading this:

https://charlesneimog.github.io/pd4web/js/

In this case, to send a float from `js` to PureData you need to use `Pd4Web.sendFloat("myreceiver", 5)`, from Pd, you will receive this number from the object `[r myreceiver]`.

----

To send, for example, a float from Pd to Js you need the function:

``` js
Pd4Web.onFloatReceived("myfloat", function (f) { 
    console.log("Received " + f)
});
```

This function will be executed always you send something to object `[s myfloat]` (on Pd side).

**Reply by luismrguimaraes** [reply]
Thanks for the quick reply! I'm having a lot of trouble creating and using Pd4Web outside the `index.html` in `WebPatch/`. I either get the error `Uncaught ReferenceError: SharedArrayBuffer is not defined` if I use a similar `index.html` to the one that is auto generated in `WebPatch/`, or `Uncaught TypeError: window.document.currentScript is null` if I simply add the imports (`WebPatch/pd4web.js` and `WebPatch/pd4web.threads.js`).

The ability to use pd-else would be wonderful, so I'd be delighted if this worked :)

**Reply by charlesneimog** [reply]
This is because you are not using a server with Cross-origin isolation, required to use SharedArrayBuffer, needed to use pd4web.

---
About `Uncaught ReferenceError: SharedArrayBuffer is not defined`

To make this work:

1. Activate the Cross-origin isolation in your server **OR**
2. Include the `pd4web.threads.js` in your index.html

IMPORTANT: note that, to make sure that pd4web.threads.js will in fact activate Cross-Origin Isolation you need to make an index.html with redirect to another folder (at least this was the solution I found). Because this pd4web creates a index.html that redirect to WebPatch folder.

---
About `Uncaught TypeError: window.document.currentScript is null`.

1) What server are you using?
2) Is this comming from `pd4web.threads.js`? 
3) I never saw this error, would be great to get a minimal way to reproduce it (maybe this is related with the IMPORTANT message above).

Check this https://github.com/charlesneimog/Pandemonio/, I guess that the problem is that you don't have: 

``` js
<!doctype html>
<html lang="en-us">
  <body>
    <script>
      window.location.href = "WebPatch/index.html"; /* whatever your folder */
    </script>
  </body>
</html>
```

----
About else library. 

You can use else library. Just require that you use the object using `else/` preffix or use `declare -lib else` in your patch.

**Reply by luismrguimaraes** [reply]
Thanks for the detailed reply! 

I do not want to use `window.location.href = "WebPatch/index.html"` because I want my application window to run instead, and I only want Pd to run in the background for the audio. I'm trying to get Pd working from an `index.html` in a parent folder, and am now able to read `pd4web version 2.0.6` from the browser terminal after adding Cross Origin Isolation, but get an error message containing `GET http://localhost:5173/pd4web.gui.js`. I'm fine without the GUI, but after I click the audio icon, "WebAudio worklet thread" fails to initialize. I believe `pd4web.js` assumes we are using it inside the `WebPatch/` folder and that's why I'm having these issues.

**Reply by charlesneimog** [reply]
pd4web.gui.js is required from the compiled code, if you don't want to use gui you need to pass the `-nogui` option from the Python compiler.

Compile using: `pd4web yourpatch.pd -nogui`.

**Reply by luismrguimaraes** [reply]
I don't mind that the GUI loads, what I really wanted to get working was basically the `index.html` that is auto-generated by pd4web, but at the **root** of my project. So, with `<head>` like this:
```html
<head>
    <title>Pd4Web</title>
    <meta charset="UTF-8" />
    <script src="./src/Audio/WebPatch/pd4web.threads.js"></script>
    <!-- Just necessary if you are using Github Pages -->
    <script src="./src/Audio/WebPatch/pd4web.js"></script>
    (...)
</head>
```
My project has a folder structure like this 
<img width="469" alt="image" src="https://github.com/user-attachments/assets/d7c54822-98ae-4783-9faa-b67d41ca372a">


But when I load `http://localhost:5173/index.html` I get those errors I mentioned above (error finding `pd4web.gui.js` and error initializing the "WebAudio worklet thread") but Pure Data works (only) when I access `http://localhost:5173/src/Audio/WebPatch/index.html`.

I don't mean to sound rude, but I see a lot of potential in your project and feel like this should work. If I am making something unclear, please let me know. I can even make a video explaining what I'm trying to do (because I **really** want this to work).

**Reply by charlesneimog** [reply]
Now I understand the issue and what you want to achieve.

I change minor things to fit to your use case, you can test it reinstalling the `pd4web` module installing it via 

`pip install git+https://github.com/charlesneimog/pd4web`

let me know if it works!

EDIT: It will not work because of `pd4web.data`.

If this happens, try to open `pd4web.js`, search by `pd4web.data`, and change it to where is the location of the `pd4web.data` (Audio/WebPatch/pd4web.data).

**Reply by charlesneimog** [reply]
Read about this, This will solve the `.data` problem

https://emscripten.org/docs/porting/files/packaging_files.html#changing-the-data-file-location

**Reply by luismrguimaraes** [reply]
I managed to get Pd working by moving all the relevant files in `WebPatch/` to the root 
<img width="470" alt="image" src="https://github.com/user-attachments/assets/4f31dc70-5bf7-4137-8998-56ebc3bf4c1c">

But I'll try this tomorrow so that I can have a cleaner root. Thanks for the updates!

**Reply by luismrguimaraes** [reply]
Hello again, I've just tried this, but I don't understand what I should change in `pd4web.js`. The paths for `pd4web.data` are all already absolute paths like: `var PACKAGE_NAME = '/Users/<username>/Documents/Code/<project-folder>/src/Audio/WebPatch/pd4web.data;`.

The errors I'm getting are a pop-up with "WebAudio worklet thread initialization failed!" and this:
<img width="1790" alt="image" src="https://github.com/user-attachments/assets/6ae4d111-fb3b-46f9-a13d-28a500dcab57">
These are the same as before the update.

**Reply by charlesneimog** [reply]
you must change the `REMOTE_PACKAGE_BASE`. In your case it will be something like:

```js
var REMOTE_PACKAGE_BASE = 'src/Audio/pd4web.data';
```

Obs.: I had not tested it myself!

**Reply by luismrguimaraes** [reply]
I'm getting the same error after changing that variable in `pd4web.js`. 

Fortunately, by moving the `WebPatch/` contents to the root of my project, I have Pd working and can send messages to the patch, and I'd say this discussion is resolved ("Using JavaScript to communicate with Pure Data
"). It would definitely be useful in the future to have Pd working without having to move the `WebPatch/` contents, so I can create an issue if you want.

**Reply by charlesneimog** [reply]
Yes, changing the folder seems a bit more complicated because you need to change it from the build process. I'll try to research this, but it's not the priority right now.
